python/ray/spark/__init__.py

```diff
@@ -0,0 +1,614 @@
+
+                This handles the case when the parent is a PySpark worker process.
+                If a user cancels the PySpark job, the worker process gets killed, regardless of
+                PySpark daemon and worker reuse settings.
+                We use prctl to ensure the command process receives SIGTERM after spark job
+                cancellation.
+                The command process itself should handle SIGTERM properly.
+                This is a no-op on macOS because prctl is not supported.
+
+                Note:
+                When a pyspark job cancelled, the UDF python process are killed by signal "SIGKILL",
+                This case neither "atexit" nor signal handler can capture SIGKILL signal.
+                prctl is the only way to capture SIGKILL signal.
+                """
+                try:
+                    import ctypes
+                    import signal
+
+                    libc = ctypes.CDLL("libc.so.6")
+                    # Set the parent process death signal of the command process to SIGTERM.
+                    libc.prctl(1, signal.SIGTERM)  # PR_SET_PDEATHSIG, see prctl.h
+                except OSError as e:
+                    _worker_logger.warning(
+                        f"Setup libc.prctl PR_SET_PDEATHSIG failed, error {repr(e)}."
+                    )
+
+        else:
+            setup_sigterm_on_parent_death = None
+
+        _worker_logger.info(f"Start Ray worker, command: {' '.join(ray_worker_cmd)}")
+
+        # TODO: write to NFS mount
+        with open(
+            os.path.join(ray_log_dir, f"ray-start-worker-{task_id}.log"),
+            "w",
+            buffering=1,
+        ) as worker_log_fp:
+            exec_cmd(
+                ray_worker_cmd,
+                synchronous=True,
+                capture_output=False,
+                stream_output=False,
+                extra_env=ray_worker_extra_envs,
+                preexec_fn=setup_sigterm_on_parent_death,
+                stdout=worker_log_fp,
+                stderr=subprocess.STDOUT,
+            )
+
+        # Delete the worker temp and log directories at the conclusion of running the
+        # submitted task.
+        shutil.rmtree(ray_temp_dir, ignore_errors=True)
+        shutil.rmtree(ray_log_dir, ignore_errors=True)
+
+        # NB: Not reachable.
+        yield 0
+
+    spark_job_group_id = (
+        f"ray-cluster-job-head-{ray_head_hostname}-port-{ray_head_port}"
+    )
+
+    ray_cluster_handler = RayClusterOnSpark(
+        address=f"{ray_head_hostname}:{ray_head_port}",
+        head_proc=ray_head_proc,
+        spark_job_group_id=spark_job_group_id,
+        num_ray_workers=num_spark_tasks,
+    )
+
+    def backgroud_job_thread_fn():
+        from pyspark.sql.utils import Py4JJavaError
+
+        _spark_log4j = spark.sparkContext._jvm.org.apache.log4j.LogManager.getLogger("ray.spark")
+
+        # TODO: decide whether to expose these as configurable env variables or args
```

Comment:
Basic retry logic. 
Discussion points: 
1. Do we want this configurable? (probably yes)
2. If there are insufficient resources to start the barrier task mode, the total time to wait for a top-level exception is 10 minutes with no indication outside of the driver Log4j logs that the DagsScheduler cannot get sufficient task slots. Do we want to improve this behavior by recommending a lower spark conf setting of retry duration and max retry counts (it retries 40 times with a 15s wait between retries before finally throwing 
`org.apache.spark.scheduler.BarrierJobSlotsNumberCheckFailed`). Since the retry failure is logged as a WARN message on the driver log4j logs `22/10/21 23:03:34 WARN DAGScheduler: Barrier stage in job 27 requires 32 slots, but only 16 are available. Will retry up to 35 more times`, the main Python process has no readily available access to the contents of this. Perhaps this can be solved through documentation and recommended configuration details?

Reply:
emm,
From my side, I prefer to add the retry logic inside spark task routine, so that it can make each task do retry individually.

Reply:
On second thoughs,

I am considering using normal spark job instead of barrier mode job, the reason is:

1. Using normal spark job, spark tasks can automatically retry individually, we don't need to write additional retry logic, But, in barrier mode, if one spark task fails, it will cause all other spark tasks killed.

2. Using normal spark job, we can support failover when a spark worker physical machine crashes. (spark will try to re-schedule the spark task to other spark worker nodes)

3. Using barrier mode job, if the cluster resources does not satisfy "idle spark task slots >= argument `num_spark_task`", then the barrier job will stuck and wait until enough idle task slots available. But, this behavior is not user-friendly, on a shared spark cluster, user is hard to estimate how many idle tasks available at a time, But, if using normal spark job, it can launch job with less spark tasks, and when more task slots become available, it continues to launch tasks on new available slots, and user can see the ray cluster worker increases when more slots available.

Reply:
This sounds great. I'll remove the retry logic here and switch to a standard `mapPartitions()`. Tested it out locally to see the behavior and it looks good.

Reply:
@BenWilson2 We should also update the REP to reflect this design change.